lib/grammers-client/src/client/auth.rs

```diff
@@ -251,20 +281,128 @@ impl Client {
+                if name == "SESSION_PASSWORD_NEEDED" =>
+            {
+                let password_information = self.get_password_information().await?;
+                Ok(SignInResult::PasswordRequired {
+                    password_information,
+                })
+            }
             Err(InvocationError::Rpc(RpcError { name, .. })) if name.starts_with("PHONE_CODE_") => {
                 Err(SignInError::InvalidCode)
             }
             Err(error) => Err(SignInError::Other(error)),
         }
     }
 
+    /// Extract information needed for the two-factor authentication
+    /// It's called automatically when we get SESSION_PASSWORD_NEEDED error during sign in.
+    pub async fn get_password_information(
+        &mut self,
+    ) -> Result<tl::types::account::Password, SignInError> {
+        let request = tl::functions::account::GetPassword {};
+
+        let data: tl::enums::account::Password = match self.invoke(&request).await {
+            Ok(x) => x.into(),
+            Err(error) => return Err(SignInError::Other(error)),
+        };
+
+        let password_information: tl::types::account::Password = match data {
+            tl::enums::account::Password::Password(pass) => pass,
+        };
+
+        Ok(password_information)
+    }
+
+    /// Sign in using two-factor authentication (user password)
+    ///
+    /// [`password_information`] can be obtained from [SignInError::PasswordRequired] error after
+    /// [`sign_in`] method.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use grammers_client::{SignInError, SignInResult};
+    ///
+    /// # async fn f(mut client: grammers_client::Client) -> Result<(), Box<dyn std::error::Error>> {
+    /// # const API_ID: i32 = 0;
+    /// # const API_HASH: &str = "";
+    /// # const PHONE: &str = "";
+    /// fn get_user_password(hint: &String) -> Vec<u8> {
+    ///     unimplemented!()
+    /// }
+    ///
+    /// # let token = client.request_login_code(PHONE, API_ID, API_HASH).await?;
+    /// # let code = "";
+    ///
+    /// // ... enter phone number, request login code ...
+    ///
+    /// let user = match client.sign_in(&token, &code).await {
+    ///     Ok(SignInResult::PasswordRequired {
+    ///          password_information,
+    ///     }) => {
+    ///         let mut password = get_user_password(password_information.hint.as_ref().unwrap());
+    ///
+    ///         client
+    ///             .two_factor_auth(password, password_information)
+    ///             .await.unwrap()
+    ///     }
+    ///     Ok(SignInResult::SignedIn(user)) => user,
+    ///     Ok(_) => panic!("Sign in required"),
+    ///     Err(err) => {
+    ///         panic!("Failed to sign in as a user :(\n{}");
+    ///     }
+    /// };
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn two_factor_auth(
+        &mut self,
+        password: Vec<u8>,
+        password_information: tl::types::account::Password,
```

Comment:
The signature of `Client::sign_in` is `(token: &LoginToken, code: &str)`. I think it would make sense if the signature here was the same, that is, first the "token" and then the password itself.

I am also not to keen on requiring a `Vec<u8>` as input, since most users will have "normal" passwords that don't need anything special (I am not aware of any official client that allows you to use arbitrary bytes as your password, as it would be pretty impractical). Maybe we can take `AsRef<[u8]>` instead?